Add enableSorting prop to EuiBasicTable for default sorting of table.

Anupam-dagar opened the pull request:
Fixes: #1866

### Summary

This PR adds an `enableSorting` prop to `EuiBasicTable` component. This prop when set to true will make all columns sortable by setting `sortable` property of each column to `true`.

### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- [x] Checked in **mobile**
- [x] Checked in **IE11** and **Firefox**
- [ ] Props have proper **autodocs**
- [ ] Added **documentation** examples
- [ ] Added or updated **jest tests**
- [ ] Checked for **breaking changes** and labeled appropriately
- [ ] Checked for **accessibility** including keyboard-only and screenreader modes
- [ ] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately


As pointed out by @cchaos in the issue, this PR doesn't yet handle the case if the data in the column isn't what the sorting function accepts. I would implement it as the discussion continues over it in this PR.

kibanamachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

anishagg17 commented:
jenkins test this

cchaos commented:
Jenkins, test this

cchaos commented on src/components/basic_table/basic_table.tsx:
```diff
@@ -189,6 +189,7 @@ interface BasicTableProps<T> extends Omit<EuiTableProps, 'onChange'> {
   rowProps?: object | RowPropsCallback<T>;
   selection?: EuiTableSelectionType<T>;
   sorting?: EuiTableSortingType<T>;
+  enableSorting?: boolean;
```
There are a few more things that will need to be done when we add new props.

### 1. Add a prop docs **comment** to this so consumers understand what it does. Something like:

```suggestion
  /**
   * Attempts to enable the default sorting ability for each column
   **/
  enableSorting?: boolean;
```

EuiBasicTable is also unique in that it has it's own props docs configuration file buried in the docs site. You'll find that [file here](https://github.com/elastic/eui/blob/master/src-docs/src/views/tables/basic/props_info.js), in which you'll need to add another section for this new prop and add a description similar to the comment above.

### 2. Add a **test** for this prop in [basic_table.test.tsx](https://github.com/elastic/eui/blob/master/src/components/basic_table/basic_table.test.tsx)

You can follow the pattern by copying and pasting [this block](https://github.com/elastic/eui/blob/ca74ad31aafa7ecc163b137b99fa719b21daae46/src/components/basic_table/basic_table.test.tsx#L375-L395) and changing the configs to test the `enableSorting` prop.

### 3. Add a sentence to the table sorting section alluding to this new prop.

This section is where it talks about sorting. https://github.com/elastic/eui/blob/ca74ad31aafa7ecc163b137b99fa719b21daae46/src-docs/src/views/tables/sorting/sorting_section.js#L23-L28

I'd change it to read something like:

```js
    <p>
      The following example shows how to configure column sorting via the{' '}
      <EuiCode>sorting</EuiCode>
      property and flagging the sortable columns as{' '}
      <EuiCode>sortable: true</EuiCode>. You can also attempt to enable the default sorting behavior of all 
     columns by passing <EuiCode>enableSorting: true</EuiCode> to EuiBasicTable without the need to specify
     it for each individual column.
    </p>
```

### 4. Add a [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry

chandlerprall replied:
I think it makes more sense to move this onto the `sort` prop, and maybe rename it to e.g. `enableAllColumns`. This would help any confusion around "I added this prop but don't see any changes".

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2906/

cchaos commented:
@Anupam-dagar Have you tested this new option in the docs? I'm not seeing any change to the EuiBasicTable render (none of the table columns have the sorting functionality) when I change the prop to be true.

Anupam-dagar commented:
@cchaos Thanks for your review!
I tried this on the sorting, pagination and footer table example by adding `enableSorting={true}` to the props. I was able to see the sorting working. For sorting example, I removed `sortable: true` from all the columns to see if it is working.
With my understanding of the issue, please correct me if I am wrong, I assume that `sort` and `onchange` function will be provided and `enableSorting` will just make every column sortable.

I have attached the gif below on how I tested it.
![euia-min](https://user-images.githubusercontent.com/21174572/75254992-333a2a80-581c-11ea-9c0f-73f36e5d19b0.gif)

chandlerprall commented on src/components/basic_table/basic_table.tsx:
```diff
@@ -633,6 +635,10 @@ export class EuiBasicTable<T = any> extends Component<
 
     columns.forEach((column: EuiBasicTableColumn<T>, index: number) => {
+      if (this.props.enableSorting) {
+        (column as EuiTableFieldDataColumnType<T>).sortable = true;
```
* I'd like to see this check if the column is a `EuiTableFieldDataColumnType` (_field_ is present) or a `EuiTableComputedColumnType` (_render_ is present) before modifying the `sortable` value
* Instead of overriding any potentially existing value, this should only set to true if `sortable` isn't already defined
* This modifies the `column` object directly, which is an anti-pattern in React. Instead, this should create a new object with the modified value, e.g.

```
column = {
  ...column,
  sortable: true,
};
```

Anupam-dagar commented:
@chandlerprall I have moved the prop to `sort` prop itself. I kept it optional so that it doesn't break the current sorting implementations. I added the code block inside the `field` check, I guess `sortable` can be made `true` to `field` only columns since columns having `render` needs a function for `sortable`.

chandlerprall commented on src/components/basic_table/table_types.ts:
```diff
@@ -69,6 +69,7 @@ export interface EuiTableSortingType<T> {
   sort?: {
     field: keyof T;
     direction: Direction;
+    enableAllColumns?: boolean;
```
While testing this out, having this prop on `sort` doesn't feel right - let's move it up to the top-level `sorting`/`EuiTableSortingType` definition (next to `allowNeutralSort`). This will re-affect the documentation change in *src-docs/src/views/tables/basic/props_info.js*

Sorry for asking to move it a second time!

Anupam-dagar replied:
Thanks for this suggestion! I was also thinking of the same.

chandlerprall commented on src-docs/src/views/tables/sorting/sorting_section.js:
```diff
@@ -24,7 +24,11 @@ export const section = {
       <EuiCode>sorting</EuiCode>
       property and flagging the sortable columns as{' '}
-      <EuiCode>sortable: true</EuiCode>
+      <EuiCode>sortable: true</EuiCode>. You can also attempt to enable the
```
```suggestion
      <EuiCode>sortable: true</EuiCode>. You can also set the
```

chandlerprall commented on src-docs/src/views/tables/sorting/sorting_section.js:
```diff
@@ -24,7 +24,11 @@ export const section = {
       property and flagging the sortable columns as{' '}
-      <EuiCode>sortable: true</EuiCode>
+      <EuiCode>sortable: true</EuiCode>. You can also attempt to enable the
+      default sorting behavior of all columns by passing
```
```suggestion
      default sorting behavior of field columns by passing
```

Anupam-dagar commented:
@chandlerprall I have made the requested changes.

chandlerprall commented on src/components/basic_table/basic_table.tsx:
```diff
@@ -690,14 +690,29 @@ export class EuiBasicTable<T = any> extends Component<
+        if (
+          this.props.sorting.sort &&
+          !!this.props.sorting.enableAllColumns &&
+          !(column as EuiTableFieldDataColumnType<T>).sortable
```
Need to check explicitly for non-existance of the field, instead of falsey, so that a value of `false` can be used to disable sorting on a column.

```suggestion
          (column as EuiTableFieldDataColumnType<T>).sortable == null
```

Anupam-dagar commented:
@chandlerprall Implemented the requested changes.

cchaos commented:
Jenkins, test this

cchaos commented on src-docs/src/views/tables/sorting/sorting_section.js:
```diff
@@ -24,7 +24,11 @@ export const section = {
+      behavior of field columns by passing
+      <EuiCode>enableAllColumns: true</EuiCode> to <EuiCode>sorting</EuiCode>
+      prop of EuiBasicTable without the need to specify it for each individual
+      column.
```
```suggestion
      <EuiCode>sortable: true</EuiCode>. To enable the default sorting ability to <strong>every</strong>
      column, pass <EuiCode>enableAllColumns: true</EuiCode> to the <EuiCode>sorting</EuiCode>
      prop.
```

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2906/

Anupam-dagar commented:
@cchaos I have made the changes to make the prop work for the mobile view.

cchaos commented on src-docs/src/views/tables/sorting/sorting_section.js:
```diff
@@ -24,7 +24,11 @@ export const section = {
       <EuiCode>sorting</EuiCode>
       property and flagging the sortable columns as{' '}
-      <EuiCode>sortable: true</EuiCode>
+      <EuiCode>sortable: true</EuiCode>.To enable the default sorting ability to
```
Typo here (my fault)
```suggestion
      <EuiCode>sortable: true</EuiCode>.To enable the default sorting ability for
```

Also, when there's a line break and then element tags, it gets parsed as not having any space. You then need to add `{' '}` as you'll notice in the previous lines. I'd suggest installing a linter that can auto-fix when you save. I know it helps me a lot.

Anupam-dagar commented:
@cchaos I'll remember to not merge down my commits on my next PR's. I have fixed the typo in the sorting section.

chandlerprall commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2906/

chandlerprall commented on src-docs/src/views/tables/sorting/sorting_section.js:
```diff
@@ -24,7 +24,11 @@ export const section = {
       <EuiCode>sorting</EuiCode>
       property and flagging the sortable columns as{' '}
-      <EuiCode>sortable: true</EuiCode>
+      <EuiCode>sortable: true</EuiCode>.To enable the default sorting ability
```
One last formatting issue that I can see

```suggestion
      <EuiCode>sortable: true</EuiCode>. To enable the default sorting ability
```

Anupam-dagar commented:
@chandlerprall requested changes done. I hope the PR is now complete.
I had to commit using `-n` flag since the pre commit test was giving the following error and I guess it is unrelated to the code I have edited. It's from the combo box.
```bash
yarn run v1.21.1
$ yarn tsc --noEmit && yarn lint-es && yarn lint-sass
$ /home/anupam/github/eui/node_modules/.bin/tsc --noEmit
src/components/combo_box/combo_box.tsx:651:39 - error TS2339: Property 'copyInputStyles' does not exist on type 'AutosizeInput & HTMLDivElement'.

651         this.autoSizeInputRefInstance.copyInputStyles();
                                          ~~~~~~~~~~~~~~~


Found 1 error.
```

cchaos commented:
Jenkins, test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_2906/
